gh-105973: Ensure that `http` header parser returns RFC 7230 compliant header values

bentasker opened the pull request:
<!-- gh-issue-number: gh-105973 -->
* Issue: gh-105973
<!-- /gh-issue-number -->

bentasker committed:
GH-105973: Add regression test to verify stripping of whitespace from header values

This adds a test to check for the behaviour identified in #105973

bentasker committed:
GH-105973 Update test to test multiple header whitespace permutations

Although #105973 is promarily concerned with trailing whitespace not
being stripped, it's easy/inexpensive to test for other whitespace
behaviour (to detect any other future incompatabilities).

This test change ensures that

- Leading whitespace is trimmed from header values
- Whitespace in the middle of a header value is not removed

bentasker committed:
Add test to ensure that empty header string does not throw exceptions

bentasker committed:
Add test to ensure that multi-headers are preserved

barneygale reviewed (commented):
Would it be possible to adjust `email.policy.HTTP`, rather than applying a post-process on the headers?

bentasker committed:
Adjust email.policy to strip trailing header whitespace in some circumstances

This adds a new attribute `allow_trailing_whitespace` to `EmailPolicy` and
`Policy` and amends `header_source_parse` to consider it's state.

This currently works when `allow_trailing_whitespace` is manually
coerced to `False`, but does not appear to be working with
`http.client`'s invocations

bentasker committed:
HTTP client should use HTTP policy rather than default

bentasker committed:
Add news to Misc/NEWS.d/next/Library

bentasker committed:
Rename variable to `rstrip_whitespace`

The original name was too long and disrupted the layout of
docstrings and test definitions.

bentasker committed:
Update tests for `email`

bentasker committed:
Make `HTTP` a clone of `Compat32` to preserve/restore existing behaviour

This change will break a number of `email` tests because `Compat32` does
not currently explicitly contain (or document) certain attributes

bentasker committed:
Fix `email.policy` tests, the defaults for `HTTP` now come from `Compat32`
